### PR TITLE
Blog App - processing data in models

### PR DESCRIPTION
**Pull request by lucas-crodrigues**
This is the second part of the Blog app initiated in the first week of the Rails module in the Microverse program.

The requirements are: 

> ### Project requirements
> 
> ### Create models
> -  Create model classes for all entities as shown in the [ERD diagram](https://github.com/microverseinc/curriculum-rails/blob/main/blog-app/images/blog_app_erd.png).
> -  Set up associations between models.
> 
> ### Use models to insert data
> - Familiarize yourself with ['rails c' command](https://guides.rubyonrails.org/command_line.html#bin-rails-console).
> - Open rails console for your project.
> - Create at least one user by running the following code:
>     ```ruby
>        first_user = User.create(name: 'Tom', photo: 'https://unsplash.com/photos/F_-0BxGuVvo', bio: 'Teacher from Mexico.')
>        second_user = User.create(name: 'Lilly', photo: 'https://unsplash.com/photos/F_-0BxGuVvo', bio: 'Teacher from Poland.')
>     ```
> - Create at least 4 posts written by one of the users you created by running the following code:
>     ```ruby
>        first_post = Post.create(author: first_user, title: 'Hello', text: 'This is my first post')
>     ```
> - Create at least 6 posts comments for one of the posts you created by running the following code:
>     ```ruby
>        Comment.create(post: first_post, author: second_user, text: 'Hi Tom!' )
>     ```
> - Use other [CRUD methods](https://github.com/microverseinc/curriculum-rails/blob/main/blog-app/lessons/active_record_crud.md) to find, update and delete entities.
> 
> 
> ### Create custom methods
> Your models should include some additional methods.
> -  Users
>     - A method that returns the 3 most recent posts for a given user.
> -  Posts
>     - A method that updates the posts counter for a user.
>     - A method which returns the 5 most recent comments for a given post.
> -  Comments
>     - A method that updates the comments counter for a post.
> -  Likes
>     - A method that updates the likes counter for a post.
> - Go to `rails c` and check if your methods are working.

- [x] Create models.
- [x] Use models to insert data.
- [x] Create custom methods.
- [x] Check if methods are working.

Thank you.

### CODE REVIEW

File: app/models/user.rb [hunk 1]
```diff
@@ -0,0 +1,9 @@
+class User < ApplicationRecord
+  has_many :comments, foreign_key: 'author_id', class_name: 'Comment'
+  has_many :posts, foreign_key: 'author_id', class_name: 'Post'
+  has_many :likes, foreign_key: 'author_id', class_name: 'Like'
```
Comment (rloterh) [inline app/models/user.rb]: - [ ] OPTIONAL 
It's redundant to define `class_name` in the above associations as the `has_many` relationship already defines them. Kindly simply do something like:
```ruby
has_many :posts, foreign_key: 'author_id'
has_many :comments, foreign_key: 'author_id'
has_many :likes, foreign_key: 'author_id'
```

File: app/models/user.rb [hunk 2]
```diff
@@ -0,0 +1,9 @@
...
+
+  def last_three_posts
+    posts.last(3)
+  end
```
Comment (rloterh) [inline app/models/user.rb]: - [ ] To return the _most recent 3_ posts in some order, our fetch request implementation should be similar to:
```ruby
def recent_post
  posts.order('created_at Desc').limit(3)
end
```
Kindly ensure to improve upon the same.

File: app/models/post.rb [hunk 3]
```diff
@@ -0,0 +1,17 @@
+class Post < ApplicationRecord
+  after_save :update_posts_counter
+
+  belongs_to :author, class_name: 'User'
+  has_many :comments
+  has_many :likes
+
+  def last_five_comments
+    comments.last(5)
+  end
```
Comment (rloterh) [inline app/models/post.rb]: - [ ] To return the _most 5 recent _ comments in some order, our fetch request implementation should be similar to:
```ruby
def recent_comments
  comments.order('created_at Desc').limit(5)
end
```
Kindly ensure to improve upon the same.

**Review by rloterh — CHANGES_REQUESTED** [review]
Hi Lucas, 

You did extremely great on this project! 👏 👏 
There are however a few issues that you still need to work on to go to the next project but you are almost there! 🚦 🚦 🚦  

## Highlights 👍  
- Nice job implementing the `after_save` method after data is saved
- Professional ReadMe ⭐ 

## Required Changes ♻️
_Kindly check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

_Kindly check the comments marked **[OPTIONAL]** under the review._

**You can also consider:**
- I think you should change the header title _Authors_ to _Author_ in your ReadMe, as this project is worked on by a sole contributor. 
Implementing the above change(s) will make your ReadMe/ repository stand out even better 😉

Cheers and Happy coding!👏 💻 🍷

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: app/models/user.rb (hunk 2 above)
Comment (lucas-crodrigues) [reply app/models/user.rb]: Hello @rloterh, can you tell me why I need to change this? It works properly

**Review by zainsadaqat — APPROVED** [review]
# APPROVED!!!! 🎉 🎉 🎉 🎉 🎉 🎉
## Hey @lucas-crodrigues  😊 😃 

#### Great Job 😄 👏 You've done it. Your project is complete!  There is nothing else to say other than... it's time to merge it 👍🏻 ✅ 

# Congratulations!!! 🎉 🥳 👏 🙌 🎊 



Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**